src/auth.js

```diff
@@ -1,197 +1,161 @@
+ * // transaction to a third-party service for signing, or just do simple
+ * // signing via Keypair like it does here:
+ * function signPayloadCallback(payload) {
+ *    const signature = signer.sign(hash(payload.toXDR());
+ *    return [ signature, signer.publicKey() ];
+ * }
  *
- * @returns {xdr.SorobanAuthorizationEntry}
+ * function multiPartyAuth(
+ *    server: Server,
+ *    // assume this involves multi-party auth
+ *    tx: Transaction,
+ * ) {
+ *    return server
+ *      .simulateTransaction(tx)
+ *      .then((simResult) => {
+ *          tx.operations[0].auth.map(entry =>
+ *            authorizeEntry(
+ *              entry,
+ *              signPayloadCallback,
+ *              currentLedger + 1000,
+ *              Networks.FUTURENET);
+ *          ));
  *
- * @throws {Error} if `verify(hash(envelope), signature, publicKey)` does not
- *    pass, meaning one of the arguments was not passed or built correctly
- * @throws {TypeError} if the envelope does not hold an
- *    {@link xdr.HashIdPreimageSorobanAuthorization} instance
+ *          return server.prepareTransaction(tx, Networks.FUTURENET, simResult);
+ *      })
+ *      .then((preppedTx) => {
+ *        preppedTx.sign(source);
+ *        return server.sendTransaction(preppedTx);
+ *      });
+ * }
  */
-export function buildAuthEntry(envelope, signature, publicKey) {
-  // ensure this identity signed this envelope correctly
+export async function authorizeEntry(
+  entry,
+  signingMethod,
+  validUntil,
+  networkPassphrase = Networks.FUTURENET
+) {
+  // no-op
   if (
-    !Keypair.fromPublicKey(publicKey).verify(hash(envelope.toXDR()), signature)
+    entry.credentials().switch() !==
+    xdr.SorobanCredentialsType.sorobanCredentialsAddress()
   ) {
-    throw new Error(`signature does not match envelope or identity`);
+    return entry;
   }
 
-  if (
-    envelope.switch() !== xdr.EnvelopeType.envelopeTypeSorobanAuthorization()
-  ) {
-    throw new TypeError(
-      `expected sorobanAuthorization envelope, got ${envelope.switch().name}`
+  /** @type {xdr.SorobanAddressCredentials} */
+  const addrAuth = entry.credentials().address();
+  addrAuth.signatureExpirationLedger(validUntil);
+
+  const networkId = hash(Buffer.from(networkPassphrase));
+  const preimage = xdr.HashIdPreimage.envelopeTypeSorobanAuthorization(
+    new xdr.HashIdPreimageSorobanAuthorization({
+      networkId,
+      nonce: addrAuth.nonce(),
+      invocation: entry.rootInvocation(),
+      signatureExpirationLedger: addrAuth.signatureExpirationLedger()
+    })
+  );
+  const payload = hash(preimage.toXDR());
+
+  let signature;
+  let publicKey;
+  if (signingMethod instanceof Keypair) {
+    signature = signingMethod.sign(payload);
+    publicKey = signingMethod.publicKey();
+  } else {
+    const [sig, pk] = await signingMethod(preimage);
+    signature = Buffer.from(sig);
+    publicKey = pk;
+  }
+
+  const expectedPubkey = Address.fromScAddress(addrAuth.address()).toString();
```

Comment:
If we already know the pubkey, why require the `signingMethod` to return it? Any mismatch would be caught by the `verify()` below afaict.

The advantage of not requiring the `signingMethod` to return it, is that it lowers the barrier for implementing a custom `signingMethod`. e.g. fewer round trips to a hw wallet, potentially.

Reply:
great point, idk why I didn't realize this